numpy/core/src/multiarray/descriptor.c

```diff
@@ -1664,16 +1663,21 @@ PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
         PyErr_Clear();
         /* Now check to see if the object is registered in typeDict */
         if (typeDict != NULL) {
-            item = PyDict_GetItem(typeDict, obj);
+            PyObject *item = NULL;
 #if defined(NPY_PY3K)
-            if (!item && PyBytes_Check(obj)) {
+            if (PyBytes_Check(obj)) {
                 PyObject *tmp;
                 tmp = PyUnicode_FromEncodedObject(obj, "ascii", "strict");
                 if (tmp != NULL) {
```

Comment:
Changing this to propagate the error will change the following:

```python
np.dtype, b'\xff'
```

To raise a `UnicodeDecodeError: 'ascii' codec can't decode byte 0xff in position 0: ordinal not in range(128)` instead of the previous `TypeError`. Is that ok? This is codified in the test `test_dtype_from_bytes`. I'm happy to fix this if that is indeed the preference.

Reply:
I opened #14591 so this can be reviewed and analyzed for correctness in isolation to this change.

Reply:
@jdufresne that example segfaults on my NumPy 1.16.2, so I do not think we have to worry about it too much. The TypeError I get on master is actually nicer then a decode error. I am not sure how fine it is to set a new exception while one is already on the fly... Replacing the `goto error;` with `goto fail;` will not change the behaviour but make it clear that the error is being handled... Sorry that I did not realize that before.